[Mobile Payments] Fix Set up Tap to Pay and Manage Card Reader flows

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

**_I broke In Person Payments_** 😬

Prior to this change, the two flows for IPP settings from the new SwiftUI Payments Menu were broken.

Both their view models were created when the Payments Menu loaded, and each immediately started listening for card reader connections. Any connection of a reader which didn’t match their type would result in it being automatically disconnected.

Since both were listening, any Bluetooth readers were disconnected by the Tap to Pay flow, and the Tap to Pay reader was disconnected by the Bluetooth reader flow. This made it impossible to take any In Person Payment after opening the Payments Menu.

When previously presented from UIKit, the lifecycle was different; the flow view models weren’t created until the user opened either flow, and they were disposed of when they dismissed the flow. This meant they were only listening for readers when the flow was on screen.

To fix this in SwiftUI, we use a wrapper view. The `CardReaderSettingsFlowPresentingView` and `TapToPaySettingsFlowPresentingView` are responsible for making the view models for the flows, and only do it when their `body` function is called. This happens when the wrapper is presented (i.e. the user opens the flow) not when it’s created (i.e. when the user opens the menu.

When the wrapper is dismissed, the viewModels are disposed of.

This gives us the same lifecycle as we had in UIKit, and restores the expected functionality.

### Known issues
If you have one of reader flows open in the Payments tab, and you switch to the orders tab to take a payment there, attempts to use the other reader type for payment in the orders tab will fail with payment interrupted, as shown in the video. This is _probably_ pre-existing, but I'm checking that.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Launch the app
2. Switch to a UK or US-based WooPayments store
3. Navigate to `Menu > Payments`
4. Tap `Set up Tap to Pay on iPhone`
5. Observe that you can complete the flow and test payment

Before the fix, this would have been impossible; you'd have got the `Payment interrupted` error when the built in reader was auto-disconnected by the Manage Card Readers flow.

1. Continue by opening the Manage Card Readers flow
2. Observe that the autoconnection starts if you have a known reader, or tap Connect
3. Go through the flow
4. Observe that you can finish it and connect to a reader

This would also have been impossible, for the same reason as above.

Take some more payments via the various ways of doing it, switching between TTP and Card Reader, observe that each is disconnected automatically as expected.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/2472348/fb251c96-3789-4da4-b833-83364a941ed2



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
Fix Set up Tap to Pay and Manage Card Reader flows

Prior to this change, the two flows for IPP settings from the new SwiftUI Payments Menu were broken.

Both their view models were created when the Payments Menu loaded, and each immediately started listening for card reader connections. Any connection of a reader which didn’t match their type would result in it being automatically disconnected.

Since both were listening, any Bluetooth readers were disconnected by the Tap to Pay flow, and the Tap to Pay reader was disconnected by the Bluetooth reader flow. This made it impossible to take any In Person Payment after opening the Payments Menu.

When previously presented from UIKit, the lifecycle was different; the flow view models weren’t created until the user opened either flow, and they were disposed of when they dismissed the flow. This meant they were only listening for readers when the flow was on screen.

To fix this in SwiftUI, we use a wrapper view. The `CardReaderSettingsFlowPresentingView` and `TapToPaySettingsFlowPresentingView` are responsible for making the view models for the flows, and only do it when their `body` function is called. This happens when the wrapper is presented (i.e. the user opens the flow) not when it’s created (i.e. when the user opens the menu.

When the wrapper is dismissed, the viewModels are disposed of.

This gives us the same lifecycle as we had in UIKit, and restores the expected functionality.

joshheald committed:
Explain the fix for card reader connection conflicts

joshheald commented:
> If you have one of reader flows open in the Payments tab, and you switch to the orders tab to take a payment there, attempts to use the other reader type for payment in the orders tab will fail with payment interrupted, as shown in the video. This is _probably_ pre-existing, but I'm checking that.

It's not pre-existing; it wasn't previously possible to get into this situation, as the Manage Card Reader view hid the tab bar when it was pushed. That's not possible with SwiftUI until iOS 16, so I'm going to change it to modal presentation to get the same effect.

toupper commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsFlowPresentingView.swift:
```diff
@@ -0,0 +1,25 @@
+import SwiftUI
+import Yosemite
+
+/// This wrapper exists to ensure that the `CardReaderSettingsViewModelsOrderedList` has the same lifecycle as the
```
Nice meaningful comments!

toupper commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/PaymentSettingsFlowPresentingViewController.swift:
```diff
@@ -79,6 +79,12 @@ private extension PaymentSettingsFlowPresentingViewController {
 // MARK: - SwiftUI compatibility
 //
+
+/// N.B. this should not be used directly for the two concrete use cases it enables
```
👍

toupper reviewed (approved):
Nice bug-catching and quick fix @joshheald! LGTM, 🚢 :shipit:

joshheald committed:
Show Manage Card Reader modally

Previously, with UIKit, this was pushed with the tab bar hidden. This prevented changing tab and using the rest of the app with the connection controller active (and auto-disconnecting any tap to pay connections.)

Hiding the tab bar is not possible from SwiftUI on iOS 15, so instead we can present it as a sheet, like the Tap to Pay interface, which blocks the tab bar too.

joshheald commented:
> It's not pre-existing; it wasn't previously possible to get into this situation, as the Manage Card Reader view hid the tab bar when it was pushed. That's not possible with SwiftUI until iOS 16, so I'm going to change it to modal presentation to get the same effect.

1516dba adds the modal presentation.
